File: lib/1510.rb
$stdin.binmode

N = $stdin.gets.to_i

class MyInput
  BLOCK_SIZE = 256*1024

  def next
    if @lines.nil? || @current_index == @lines.length - 1
      @lines = next_block
      @current_index = 0
    else
      @current_index += 1
    end
    @lines[@current_index]
  end

  private

  LINE_END = "\n"

  def next_block
    if data = read_stdin(BLOCK_SIZE)
      unless data[-1] == LINE_END
        loop do
          c = read_stdin 1
          data << c if c
          break if c.nil? || c == LINE_END
        end
      end
      data.split(/\s+/)
    end
  end

  def read_stdin(bytes)
    $stdin.read(bytes).tap{|data| @eof = data.nil? } unless @eof
  end

end

input = MyInput.new
candidate = nil
power = 0

N.times do
  cur = input.next
  if power == 0
    candidate = cur
    power = 1
  elsif candidate == cur
    power += 1
  else
    power -= 1
  end
end

puts candidate

File: lib/1510.scala
object Pr1510 {
  def main(args: Array[String]) = {

    def majority(n: Int) = {
      def iter(i: Int, candidate: Int, power: Int): Int = {
        if (i == n) candidate
        else {
          val cur = readInt
          if (power == 0) iter(i+1, cur, 1)
          else if (cur == candidate) iter(i+1, candidate, power+1)
          else iter(i+1, candidate, power-1)
        }
      }
      iter(1, readInt, 1)
    }

    println( majority(readInt) )

  }
}